### PR TITLE
armsr: armv7: add STMicroelectronics boards

### PR DESCRIPTION
**Pull request by jetm**
Add missing STMicroelectronics ARM SystemReady certificated board.

Kernel configuration added from https://github.com/STMicroelectronics/device-stm-kernel and adjusted for Linux v6.1.

### CODE REVIEW

**Comment by mcbridematt** [comment]
Many thanks for doing this! I've been curious about these ST SoCs for a while now.

The config patch selects a few options already selected by either the armsr generic config or the included modules, can you check and remove ones that are already defined elsewhere?

If there is a regression in the future, it will make it easier to know what specifically the ST SoC needs to work.

**Comment by jetm** [comment]
> Many thanks for doing this! I've been curious about these ST SoCs for a while now.

Thank you for taking a look.
 
> The config patch selects a few options already selected by either the armsr generic config or the included modules, can you check and remove ones that are already defined elsewhere?

Done. I have removed the common between generic and `armv7`:
```
$ comm -12 <(sort target/linux/armsr/config-6.1) <(sort target/linux/armsr/armv7/config-6.1)
CONFIG_CRYPTO_LIB_BLAKE2S_GENERIC=y                                                                                                                                                                                                                           
CONFIG_HZ_PERIODIC=y                                                                                                                                                                                                                                          
# CONFIG_I2C_SLAVE_TESTUNIT is not set                                                                                                                                                                                                                        
CONFIG_SERIAL_OF_PLATFORM=y
```

> If there is a regression in the future, it will make it easier to know what specifically the ST SoC needs to work.

Yes, it makes sense. Thanks!

I have another commit in queue to move other common configs between `armv7` and `armv8` to the generic one:
```
$ comm -12 <(sort target/linux/armsr/armv8/config-6.1) <(sort target/linux/armsr/armv7/config-6.1)
CONFIG_CRYPTO_SHA1=y
CONFIG_DMA_SHARED_BUFFER=y
CONFIG_GPIO_GENERIC_PLATFORM=y
CONFIG_GPIO_GENERIC=y
CONFIG_I2C_BOARDINFO=y
CONFIG_I2C=y
CONFIG_MFD_CORE=y
CONFIG_MMC_ARMMMCI=y
CONFIG_RTC_I2C_AND_SPI=y
CONFIG_USB_OHCI_HCD_PLATFORM=y
CONFIG_USB_OHCI_HCD=y
```

File: target/linux/armsr/armv7/config-6.1 [hunk 1]
```diff
@@ -1,17 +1,26 @@
...
 CONFIG_ARCH_MULTIPLATFORM=y
 CONFIG_ARCH_MULTI_V6_V7=y
 CONFIG_ARCH_MULTI_V7=y
-CONFIG_ARCH_MMAP_RND_BITS=8
 CONFIG_ARCH_NR_GPIO=0
 CONFIG_ARCH_OPTIONAL_KERNEL_RWX=y
 CONFIG_ARCH_OPTIONAL_KERNEL_RWX_DEFAULT=y
 CONFIG_ARCH_SELECT_MEMORY_MODEL=y
 CONFIG_ARCH_VIRT=y
+CONFIG_ARCH_STI=y
```
Comment (linusw) [inline target/linux/armsr/armv7/config-6.1:14]: Why are you activating ARCH_STI? This is used in set-top boxed and Starlink, none used by OpenWrt I think.

File: target/linux/armsr/armv7/config-6.1 [hunk 2]
```diff
@@ -42,36 +63,216 @@ CONFIG_CPU_SPECTRE=y
...
 CONFIG_OLD_SIGACTION=y
 CONFIG_OLD_SIGSUSPEND3=y
+CONFIG_OPTEE=y
 CONFIG_OUTER_CACHE=y
 CONFIG_OUTER_CACHE_SYNC=y
 CONFIG_PAGE_OFFSET=0xC0000000
+CONFIG_PCS_XPCS=y
 CONFIG_PERF_USE_VMALLOC=y
+CONFIG_PINCTRL=y
+CONFIG_PINCTRL_ST=y
```
Comment (linusw) [inline target/linux/armsr/armv7/config-6.1:148]: This is also related to ARCH_STI which I think you don't want.

File: target/linux/armsr/armv7/config-6.1 (hunk 1 above)
Comment (linusw) [inline target/linux/armsr/armv7/config-6.1:14]: Why are you activating ARCH_STI? This is used in set-top boxed and Starlink, none used by OpenWrt I think.

**Review by linusw — CHANGES_REQUESTED** [review]
Some ARCH_STI noise needs to be dropped, I think you only want ARCH_STM32?

File: target/linux/armsr/armv7/config-6.1 (hunk 2 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1]: Yes, the certificated board is a STM32. I have removed those. Thanks!

File: target/linux/armsr/armv7/config-6.1 (hunk 1 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1]: Removed. Thanks!

File: target/linux/armsr/armv7/config-6.1 [hunk 3]
```diff
@@ -1,17 +1,25 @@
...
 CONFIG_ARCH_OPTIONAL_KERNEL_RWX=y
 CONFIG_ARCH_OPTIONAL_KERNEL_RWX_DEFAULT=y
 CONFIG_ARCH_SELECT_MEMORY_MODEL=y
 CONFIG_ARCH_VIRT=y
+CONFIG_ARCH_STM32=y
+CONFIG_ARM_GLOBAL_TIMER=y
+CONFIG_ARM_GT_INITIAL_PRESCALER_VAL=1
 CONFIG_ARM=y
 CONFIG_ARM_CPU_SUSPEND=y
+CONFIG_ARM_ERRATA_754322=y
```
Comment (linusw) [inline target/linux/armsr/armv7/config-6.1:19]: Why are you selecteding this errata? It is selected by STI but not STM32.

File: target/linux/armsr/armv7/config-6.1 [hunk 4]
```diff
@@ -1,17 +1,25 @@
...
+CONFIG_ARM_ERRATA_764369=y
```
Comment (linusw) [inline target/linux/armsr/armv7/config-6.1:20]: Dito

File: target/linux/armsr/armv7/config-6.1 [hunk 5]
```diff
@@ -1,17 +1,25 @@
...
+CONFIG_ARM_ERRATA_775420=y
```
Comment (linusw) [inline target/linux/armsr/armv7/config-6.1:21]: Dito

File: target/linux/armsr/armv7/config-6.1 [hunk 6]
```diff
@@ -21,13 +29,25 @@ CONFIG_ARM_LPAE=y
...
 CONFIG_ARM_PSCI=y
+CONFIG_ARM_SMC_MBOX=y
+CONFIG_ARM_STM32_CPUIDLE=y
 CONFIG_ARM_THUMB=y
 CONFIG_ARM_UNWIND=y
 CONFIG_ARM_VIRT_EXT=y
 CONFIG_AUTO_ZRELADDR=y
 CONFIG_BINFMT_FLAT_ARGVP_ENVP_ON_STACK=y
 CONFIG_CACHE_L2X0=y
+CONFIG_CLKSRC_ARM_GLOBAL_TIMER_SCHED_CLOCK=y
```
Comment (linusw) [inline target/linux/armsr/armv7/config-6.1:40]: Why are you selecting this? I don't thing STM32 needs this or has this timer, but please explain I might be wrong.

File: target/linux/armsr/armv7/config-6.1 [hunk 7]
```diff
@@ -21,13 +29,25 @@ CONFIG_ARM_LPAE=y
...
+CONFIG_CLKSRC_MMIO=y
+CONFIG_CLKSRC_STM32=y
+CONFIG_CLKSRC_ST_LPC=y
```
Comment (linusw) [inline target/linux/armsr/armv7/config-6.1:43]: Drop this, only used by STI

File: target/linux/armsr/armv7/config-6.1 [hunk 8]
```diff
@@ -42,36 +62,205 @@ CONFIG_CPU_SPECTRE=y
...
+# CONFIG_KEYBOARD_ST_KEYSCAN is not set
+CONFIG_GPIO_GENERIC=y
+CONFIG_GPIO_GENERIC_PLATFORM=y
+CONFIG_GPIO_SYSCON=y
+CONFIG_GPIO_SYSFS=y
 CONFIG_HARDEN_BRANCH_PREDICTOR=y
 CONFIG_HAVE_SMP=y
+CONFIG_HW_RANDOM=y
+CONFIG_HW_RANDOM_OPTEE=y
+CONFIG_HW_RANDOM_ST=y
```
Comment (linusw) [inline target/linux/armsr/armv7/config-6.1:93]: I think this one is for STI (not sure but I think so?)

File: target/linux/armsr/armv7/config-6.1 [hunk 9]
```diff
@@ -42,36 +62,205 @@ CONFIG_CPU_SPECTRE=y
...
 CONFIG_MIGHT_HAVE_CACHE_L2X0=y
+CONFIG_MMC=y
+CONFIG_MMC_ARMMMCI=y
+CONFIG_MMC_CQHCI=y
+CONFIG_MMC_RICOH_MMC=y
+CONFIG_MMC_SDHCI=y
+CONFIG_MMC_SDHCI_IO_ACCESSORS=y
+CONFIG_MMC_SDHCI_PCI=y
+CONFIG_MMC_SDHCI_PLTFM=y
+CONFIG_MMC_SDHCI_ST=y
```
Comment (linusw) [inline target/linux/armsr/armv7/config-6.1:129]: I think you can drop all the SDHCI config options, AFAIK they are only used on STI.

File: target/linux/armsr/armv7/config-6.1 [hunk 10]
```diff
@@ -42,36 +62,205 @@ CONFIG_CPU_SPECTRE=y
...
+CONFIG_OPTEE=y
 CONFIG_OUTER_CACHE=y
 CONFIG_OUTER_CACHE_SYNC=y
 CONFIG_PAGE_OFFSET=0xC0000000
+CONFIG_PCS_XPCS=y
 CONFIG_PERF_USE_VMALLOC=y
+CONFIG_PINCTRL_STM32=y
+CONFIG_PINCTRL_STM32MP135=y
+CONFIG_PINCTRL_STM32MP157=y
+CONFIG_PL310_ERRATA_753970=y
```
Comment (linusw) [inline target/linux/armsr/armv7/config-6.1:149]: Drop this, only used by STI

File: target/linux/armsr/armv7/config-6.1 [hunk 11]
```diff
@@ -42,36 +62,205 @@ CONFIG_CPU_SPECTRE=y
...
+CONFIG_PL310_ERRATA_769419=y
```
Comment (linusw) [inline target/linux/armsr/armv7/config-6.1:150]: Dito

File: target/linux/armsr/armv7/config-6.1 [hunk 12]
```diff
@@ -42,36 +62,205 @@ CONFIG_CPU_SPECTRE=y
...
+CONFIG_SPI_MEM=y
+CONFIG_SPI_SPIDEV=y
+CONFIG_SPI_STM32=y
+CONFIG_SPI_STM32_QSPI=y
+CONFIG_SRAM=y
+CONFIG_SRAM_EXEC=y
+CONFIG_STM32_EXTI=y
+CONFIG_STM32_THERMAL=y
+CONFIG_STM32_WATCHDOG=y
+CONFIG_ST_IRQCHIP=y
```
Comment (linusw) [inline target/linux/armsr/armv7/config-6.1:218]: Drop this, only for STI.

File: target/linux/armsr/armv7/config-6.1 [hunk 13]
```diff
@@ -42,36 +62,205 @@ CONFIG_CPU_SPECTRE=y
...
+# CONFIG_ST_LPC_WATCHDOG is not set
+CONFIG_ST_THERMAL=y
```
Comment (linusw) [inline target/linux/armsr/armv7/config-6.1:220]: Drop this, only STI, but why is CONFIG_STM32_THERMAL not selected? I can't find it!

File: target/linux/armsr/armv7/config-6.1 [hunk 14]
```diff
@@ -42,36 +62,205 @@ CONFIG_CPU_SPECTRE=y
...
+CONFIG_STM32_DMA=y
+CONFIG_STM32_DMAMUX=y
+CONFIG_STM32_FMC2_EBI=y
+CONFIG_STM32_IPCC=y
+CONFIG_STM32_LPTIMER_CNT=y
+CONFIG_STM32_MDMA=y
+CONFIG_STM32_RPROC=y
+CONFIG_STM32_TIMER_CNT=y
+CONFIG_STMMAC_ETH=y
+CONFIG_ST_THERMAL_MEMMAP=y
```
Comment (linusw) [inline target/linux/armsr/armv7/config-6.1:236]: Drop this, only STI.

**Review by linusw — CHANGES_REQUESTED** [review]
I found a bunch of other STI-related configs when I looked closer, please check these and see if you really need them or they can be dropped. The erratas have sometimes severe performance impact so check those really close.

File: target/linux/armsr/armv7/config-6.1 (hunk 4 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1:20]: Done

File: target/linux/armsr/armv7/config-6.1 (hunk 3 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1:19]: Removed. Thanks!

File: target/linux/armsr/armv7/config-6.1 (hunk 5 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1:21]: Done

File: target/linux/armsr/armv7/config-6.1 (hunk 6 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1:40]: Removed

File: target/linux/armsr/armv7/config-6.1 (hunk 8 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1:93]: Yes, you're right. Thanks!

File: target/linux/armsr/armv7/config-6.1 (hunk 7 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1:43]: Done. Thanks

File: target/linux/armsr/armv7/config-6.1 (hunk 9 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1:129]: Remove them. Thanks

File: target/linux/armsr/armv7/config-6.1 (hunk 11 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1:150]: Remove them. Thanks

File: target/linux/armsr/armv7/config-6.1 (hunk 13 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1:220]: `CONFIG_STM32_THERMAL` is selected, just before `CONFIG_STM32_WATCHDOG`. Removed the rest. Thanks

File: target/linux/armsr/armv7/config-6.1 (hunk 10 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1:149]: Done

File: target/linux/armsr/armv7/config-6.1 (hunk 14 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1:236]: Done

File: target/linux/armsr/armv7/config-6.1 (hunk 12 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1:218]: Done

**Review by linusw — APPROVED** [review]
Looks good to me!

File: target/linux/armsr/armv7/config-6.1 [hunk 15]
```diff
@@ -42,36 +55,188 @@ CONFIG_CPU_SPECTRE=y
...
+CONFIG_I2C_DEMUX_PINCTRL=y
+CONFIG_I2C_MUX=y
+CONFIG_I2C_MUX_PCA954x=y
+CONFIG_I2C_MUX_PINCTRL=y
+CONFIG_I2C_STM32F7=y
+CONFIG_IIO_STM32_LPTIMER_TRIGGER=y
+# CONFIG_INTEL_QEP is not set
+# CONFIG_INTERRUPT_CNT is not set
+# CONFIG_MICROCHIP_TCB_CAPTURE is not set
+CONFIG_JFFS2_FS=y
```
Comment (hauke) [inline target/linux/armsr/armv7/config-6.1:103]: Why do you need `CONFIG_JFFS2_FS`?

File: target/linux/armsr/armv7/config-6.1 [hunk 16]
```diff
@@ -42,36 +55,188 @@ CONFIG_CPU_SPECTRE=y
...
+CONFIG_CRYPTO_DEV_STM32_CRC=y
+CONFIG_CRYPTO_DEV_STM32_CRYP=y
+CONFIG_CRYPTO_DEV_STM32_HASH=y
+CONFIG_CRYPTO_SHA1=y
 CONFIG_DEBUG_LL_INCLUDE="mach/debug-macro.S"
 CONFIG_DMA_OPS=y
+CONFIG_DMA_VIRTUAL_CHANNELS=y
+CONFIG_DMA_SHARED_BUFFER=y
+CONFIG_DRM_STM=y
+CONFIG_DRM_STM_DSI=y
```
Comment (hauke) [inline target/linux/armsr/armv7/config-6.1:67]: Do you really need DRM support in the kernel?
Is this needed at all in OpenWrt? Why not put it into a kernel module?

File: target/linux/armsr/armv7/config-6.1 [hunk 17]
```diff
@@ -42,36 +55,188 @@ CONFIG_CPU_SPECTRE=y
...
+CONFIG_RTC_DRV_STM32=y
+CONFIG_RTC_I2C_AND_SPI=y
 CONFIG_RTC_MC146818_LIB=y
-CONFIG_SERIAL_OF_PLATFORM=y
+CONFIG_SD_ADC_MODULATOR=y
+CONFIG_SENSORS_IIO_HWMON=y
+CONFIG_SERIAL_STM32=y
+CONFIG_SERIAL_STM32_CONSOLE=y
 CONFIG_SMP_ON_UP=y
+CONFIG_SND=y
```
Comment (hauke) [inline target/linux/armsr/armv7/config-6.1:169]: I would prefer if sound is in a kmod package

Comment (jetm) [reply target/linux/armsr/armv7/config-6.1]: I have removed the SND support, as it's not required for now. In the future, the already defined `kmod-sound-core` could be added as kmod package.

File: target/linux/armsr/armv7/config-6.1 (hunk 16 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1]: There is no requirement for now. I have removed the DRM support. Thanks

File: target/linux/armsr/armv7/config-6.1 (hunk 15 above)
Comment (jetm) [reply target/linux/armsr/armv7/config-6.1]: It's not required. I have disabled `CONFIG_JFFS2_FS`. Thanks!

**Comment by hauke** [comment]
Please refresh the configuration using this command: `make kernel_oldconfig CONFIG_TARGET=subtarget

**Comment by jetm** [comment]
Closing this PR in favor of https://github.com/openwrt/openwrt/pull/13684 as armv7, it's out of scope for SystemReady Certifications boards.

**Pull request closed without merging**